ospec.js

```diff
@@ -278,116 +547,176 @@ else window.o = m()
+	function succeed(i, message, error) {
+		var result = results[i]
 		result.pass = true
-		result.context = subjects.join(" > ")
 		result.message = message
+		// for notSatisfies. Use the testError for other passing assertions
+		if (error != null) result.error = error
 	}
-	function fail(result, message, error) {
+
+	function fail(i, message, error) {
+		var result = results[i]
 		result.pass = false
-		result.context = subjects.join(" > ")
 		result.message = message
 		result.error = error != null ? error : ensureStackTrace(new Error)
 	}
+
 	function serialize(value) {
 		if (hasProcess) return require("util").inspect(value) // eslint-disable-line global-require
 		if (value === null || (typeof value === "object" && !(value instanceof Array)) || typeof value === "number") return String(value)
 		else if (typeof value === "function") return value.name || "<anonymous function>"
 		try {return JSON.stringify(value)} catch (e) {return String(value)}
 	}
-	function noTimeoutRightNow() {
-		throw new Error("o.timeout must be called snchronously from within a test definition or a hook.")
+
+	// o.spy is functionally equivalent to this:
+	// the extra complexity comes from compatibility issues
+	// in ES5 environments where you can't overwrite fn.length
+
+	// o.spy = function(fn) {
+	// 	var spy = function() {
+	// 		spy.this = this
+	// 		spy.args = [].slice.call(arguments)
+	// 		spy.calls.push({this: this, args: spy.args})
+	// 		spy.callCount++
+
+	// 		if (fn) return fn.apply(this, arguments)
+	// 	}
+	// 	if (fn)
+	// 		Object.defineProperties(spy, {
+	// 			length: {value: fn.length},
+	// 			name: {value: fn.name}
+	// 		})
+	// 	spy.args = []
+	// 	spy.calls = []
+	// 	spy.callCount = 0
+	// 	return spy
+	// }
+
+	var spyFactoryCache = Object.create(null)
+
+	function makeSpyFactory(name, length) {
+		if (spyFactoryCache[name] == null) spyFactoryCache[name] = []
+		var args = Array.apply(null, {length: length}).map(
+			function(_, i) {return "_" + i}
+		).join(", ");
+		var code =
+			"'use strict';" +
+			"var spy = (0, function " + name + "(" + args + ") {" +
+			"   return helper(this, [].slice.call(arguments), fn, spy)" +
+			"});" +
+			"return spy"
+
+		return spyFactoryCache[name][length] = new Function("fn", "helper", code)
+	}
+
+	function getOrMakeSpyFactory(name, length) {
+		return spyFactoryCache[name] && spyFactoryCache[name][length] || makeSpyFactory(name, length)
+	}
+
+	function spyHelper(self, args, fn, spy) {
+		spy.this = self
+		spy.args = args
+		spy.calls.push({this: self, args: args})
+		spy.callCount++
+
+		if (fn) return fn.apply(self, args)
+	}
+
+	var supportsFunctionMutations = false;
+	// eslint-disable-next-line no-empty, no-implicit-coercion
+	try {supportsFunctionMutations = !!Object.defineProperties(function(){}, {name: {value: "a"},length: {value: 1}})} catch(_){}
+
+	var supportsEval = false
+	// eslint-disable-next-line no-eval, no-empty
+	try {eval("supportsEval = true")} catch(e){}
```

Comment:
Better:

```suggestion
	// eslint-disable-next-line no-new-func, no-empty
	try {supportsEval = Function("return true")()} catch(e){}
```

That way, we're not potentially interfering with engines' ability to optimize everything else.